axum/src/serve.rs

```diff
@@ -18,12 +18,59 @@ use hyper_util::rt::{TokioExecutor, TokioIo};
 #[cfg(any(feature = "http1", feature = "http2"))]
 use hyper_util::{server::conn::auto::Builder, service::TowerToHyperService};
 use tokio::{
+    io::{AsyncRead, AsyncWrite},
     net::{TcpListener, TcpStream},
     sync::watch,
 };
 use tower::ServiceExt as _;
 use tower_service::Service;
 
+/// Types that can listen for connections.
+pub trait Listener: Send + 'static {
+    /// The listener's IO type.
+    type Io: AsyncRead + AsyncWrite + Unpin + Send + 'static;
+
+    /// The listener's address type.
+    type Addr: Send;
+
+    /// Accept a new incoming connection to this listener
+    fn accept(&mut self) -> impl Future<Output = io::Result<(Self::Io, Self::Addr)>> + Send;
+
+    /// Returns the local address that this listener is bound to.
+    fn local_addr(&self) -> io::Result<Self::Addr>;
+}
+
+impl Listener for TcpListener {
+    type Io = TcpStream;
+    type Addr = std::net::SocketAddr;
+
+    #[inline]
+    async fn accept(&mut self) -> io::Result<(Self::Io, Self::Addr)> {
```

Comment:
Currently, we ignore these errors. Or more precisely, we ignore "connection errors" and log and sleep on all other errors.

I think we could just have a return type of `(Self::Io, Self::Addr)` here and let the underlying method do whatever it wants to handle transient accept errors, log them, sleep, panic, or anything else.

Or we could let it return a `Result` but then I think those errors should be fatal and we'd just propagate them. And in that case I'd also consider whether to have an associated type for the error type instead of forcing `std::io::Error`.